### ISSUE TITLE
urlsplit manufactures hostnames because it strips off tabs before validating them

### ISSUE THREAD

**Issue by gh-andre** [python/cpython#122761]
# Bug report

### Bug description:

```python
import urllib.parse

# prints "abcxyz.test"
print(urllib.parse.urlsplit("http://abc\txyz.test/").netloc)
```
Current `urlsplit` is implemented according to this spec:

https://url.spec.whatwg.org/#concept-basic-url-parser

The spec does say in item 3 to strip tabs, but I believe there's a bug in the specification (perhaps they wanted to say leading/trailing whitespace) because the item 7 in [host parsing](https://url.spec.whatwg.org/#concept-host-parser) says

> If asciiDomain contains a [forbidden domain code point](https://url.spec.whatwg.org/#forbidden-domain-code-point), [domain-invalid-code-point](https://url.spec.whatwg.org/#domain-invalid-code-point) [validation error](https://url.spec.whatwg.org/#validation-error), return failure.

, and tab is listed as a "forbidden domain code point". If tabs are stripped from the entire input before any other work is done, checking for tabs in host names wouldn't make much sense.

I created a bug in the specification project, so maybe they will provide some guidance later on.

https://github.com/whatwg/url/issues/829

### CPython versions tested on:

3.10

### Operating systems tested on:

Linux, Windows

**Comment by gh-andre** [comment]
It was pointed out to me in the referenced issue that the spec already has the provision for tracking tabs and newlines as non-terminating validation errors. Here's related items 2 and 3 from the spec:

https://url.spec.whatwg.org/#concept-basic-url-parser

> If input contains any [ASCII tab or newline](https://infra.spec.whatwg.org/#ascii-tab-or-newline), [invalid-URL-unit](https://url.spec.whatwg.org/#invalid-url-unit) [validation error](https://url.spec.whatwg.org/#validation-error).

> Remove all [ASCII tab or newline](https://infra.spec.whatwg.org/#ascii-tab-or-newline) from input.

Validation errors are indeed not supposed to terminate parsing, but given that in many contexts silently stripping tabs and newlines manufactures modifies domains, my suggestion would be to provide an optional parameter in `urlsplit`, which would fail early if tabs, newlines and non-URL-code-points are encountered in the URLs.

In other words, allow callers to request `urlsplit` to fail on validation errors in this step.

This will keep the existing behavior exactly as-is, but will allow callers operating in contexts where URLs are expected to be syntactically valid to fail early and not introduce modified domains and other URL components with stripped tabs and newlines or spend CPU cycles on fully parsing malformed URLs.
